src/ua/main.cpp

```diff
@@ -529,6 +538,63 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    map<string, JSON> fileDescriptions;
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+
+      // Cache file descriptions so we only have to do once per file,
+      // not once per chunk.
+      if (fileDescriptions.find(c->fileID) == fileDescriptions.end())
+          fileDescriptions[c->fileID] = fileDescribe(c->fileID);
+
+      if (!is_chunk_complete(c, fileDescriptions[c->fileID])) {
```

Comment:
Assuming that I've identified some correctness issues below, we should ensure this code is actually tested. Let's force this branch to be taken. I suggest (for testing only) change this line to take the branch 10% of the time, and then upload something that will be many parts. Perhaps make NUM_CHUNK_CHECKS very big to test that we converge to a completely finished state.